internal/pkg/call/call.go

```diff
@@ -0,0 +1,55 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package call contains functionality for determining whether a function
+// call refers to a given referrer.
+package call
+
+import (
+	"golang.org/x/tools/go/ssa"
+)
+
+// A Call is a wrapper for *ssa.Call instructions, allowing Variadic and
+// non-Variadic calls to be handled polymorphically
+type Call interface {
+	// ReferredBy determines if the call is referred to by the referrer
+	ReferredBy(r Referrer) bool
+}
+
+// A Referrer is something that can be queried to determine whether it refers to
+// a given ssa Node.
+type Referrer interface {
+	RefersTo(node ssa.Node) bool
+}
+
+// A RegularCall is a wrapper around a non-variadic ssa Call.
+type RegularCall struct {
+	*ssa.Call
+}
+
+// Regular creates a RegularCall from an ssa Call.
+func Regular(c *ssa.Call) *RegularCall {
+	return &RegularCall{c}
+}
+
+// ReferredBy determines if the supplied referrer refers one of the Call's arguments.
+func (c *RegularCall) ReferredBy(r Referrer) bool {
+	for _, a := range c.Call.Call.Args {
+		if r.RefersTo(a.(ssa.Node)) {
+			return true
+		}
+	}
+
+	return false
+}
```

Comment:
What do you think about the pattern we used in testing the source pkg?
Concretely, `source` implements the `Stinger` interface, and the logic there describes how a source is connected to potential sinks. 
During tests, once we find a `source`, we report it (within the analyzer) by calling its String method.
https://github.com/google/go-flow-levee/blob/master/internal/pkg/source/source_test.go#L83

In our testdata, we annotate sources with the expected string representation of a `source`.
https://github.com/google/go-flow-levee/blob/master/internal/pkg/source/testdata/src/allocation/test.go#L22

How is this helpful:
1. Tests become more declarative - simply annotate the calls under tests with the expected string representation (ex. regular call, variadic call).
2. We no longer need to collect the calls in the analyzer and have these checks that the expected number of calls was collected.

WDYT?

Reply:
(For context, while writing the tests I was using the `varargs` package as a reference.)

I think the pattern you are describing makes sense for the `source` package, but I don't think it makes sense for `call` or `varargs`, because we would need to turn them into full-blown analyzers, which I feel isn't necessary for them to do their job and therefore would make the code harder to understand. The testing code would certainly be cleaner, but the actual implementation code would be more complicated, which I think would be detrimental overall. 

I've reorganized the code in a way that I think makes it more readable. Specifically, I put the test function first, so the reader doesn't have to slog through the analyzer's definition, which is not necessary to understand the test.

Reply:
> For context, while writing the tests I was using the `varargs` package as a reference.
   
    Currently, we are inconsistent in how we organize our tests. The pattern we introduced in the source pkg came after the 
    vararg pkg was implemented. This was on my TODO list to bring consistency in how our tests are organized.

> ... because we would need to turn them into full-blown analyzers
   PTAL at my comment here: https://github.com/google/go-flow-levee/pull/34/files#r465973836

Please note that your code is well organized - this is not the issue here.
The reason why I am bringing this up is consistency, and my hope was to use the pattern we currently use in the `source` pkg.

Reply:
I see, thank you for explaining!